File: python/socketd/socketd/transport/core/listener/PipelineListener.py

from socketd.transport.core import Listener
from socketd.transport.core.Message import Message
from socketd.transport.core.Session import Session
from socketd.utils.RunUtils import RunUtils


class PipelineListener(Listener):

    def __init__(self):
        self._deque: list[Listener] = []

    def prev(self, listener: Listener) -> 'PipelineListener':
        self._deque.insert(0, listener)
        return self

    def next(self, listener: Listener) -> 'PipelineListener':
        self._deque.append(listener)
        return self

    def size(self) -> int:
        return self._deque.__len__()

    async def on_open(self, session: Session):
        for listener in self._deque:
            await RunUtils.waitTry(listener.on_open(session))

    async def on_message(self, session: Session, message: Message):
        for listener in self._deque:
            await RunUtils.waitTry(listener.on_message(session, message))

    async def on_reply(self, session: Session, message: Message):
        for listener in self._deque:
            await RunUtils.waitTry(listener.on_reply(session, message))

    async def on_send(self, session: Session, message: Message):
        for listener in self._deque:
            await RunUtils.waitTry(listener.on_send(session, message))

    async def on_close(self, session: Session):
        for listener in self._deque:
            await RunUtils.waitTry(listener.on_close(session))

    async def on_error(self, session: Session, error:Exception):
        for listener in self._deque:
            await RunUtils.waitTry(listener.on_error(session, error))

File: javascript/socketd/src/socketd/transport/core/listener/PipelineListener.ts
import type {Session} from "../Session";
import type {Message} from "../Message";
import {Listener} from "../Listener";

/**
 * 管道监听器
 *
 * @author noear
 * @since 2.0
 */
export class PipelineListener implements Listener {
    protected _deque = new Array<Listener>();

    /**
     * 前一个
     */
    prev(listener: Listener): PipelineListener {
        this._deque.unshift(listener);
        return this;
    }

    /**
     * 后一个
     */
    next(listener: Listener): PipelineListener {
        this._deque.push(listener);
        return this;
    }

    /**
     * 数量（二级监听器的数据）
     * */
    size(): number {
        return this._deque.length;
    }

    /**
     * 打开时
     *
     * @param session 会话
     */
    onOpen(session: Session) {
        for (const listener of this._deque) {
            listener.onOpen(session);
        }
    }

    /**
     * 收到消息时
     *
     * @param session 会话
     * @param message 消息
     */
    onMessage(session: Session, message: Message) {
        for (const listener of this._deque) {
            listener.onMessage(session, message);
        }
    }

    /**
     * 收到答复时
     *
     * @param session 会话
     * @param message 消息
     */
    onReply(session: Session, message: Message) {
        for (const listener of this._deque) {
            listener.onReply(session, message);
        }
    }

    /**
     * 发送消息时
     *
     * @param session 会话
     * @param message 消息
     */
    onSend(session: Session, message: Message) {
        for (const listener of this._deque) {
            listener.onSend(session, message);
        }
    }

    /**
     * 关闭时
     *
     * @param session 会话
     */
    onClose(session: Session) {
        for (const listener of this._deque) {
            listener.onClose(session);
        }
    }

    /**
     * 出错时
     *
     * @param session 会话
     * @param error   错误信息
     */
    onError(session: Session, error: Error) {
        for (const listener of this._deque) {
            listener.onError(session, error);
        }
    }
}